### ISSUE TITLE
Simplify API

### ISSUE THREAD

**Issue by kenchris** [w3c/nfc#50]
I am getting this idea that our API is too complicated and it would be nice if we could make it simpler. Maybe something like:

```
Promise<sequence<NfcAdapter>> nfc.findAdapter();

void    adapter.onread = (NfcMessageEvent)
void    adapter.onpush = ()
void    adapter.setPushData(NfcMessageData data)
Promise adapter.write(NfcMessageData data)
Promise adapter.watch(WatchOptions? options)
```

I would like to avoid the whole peer vs tag because it is not that simple in NFC. It is basically active and passive devices, plus sometimes it is a card and not a tag.... like card emulation... pff

setPushData doesn't need a promise, because users will change the push data all the time depending on what it visible in the app, and no need to wait for a promise to resolve. It doesn't matter if the data can be pushed or not, only whether it is - therefore I added the onpush.

The onread can be used for both active and passive devices. If we need to differentiate, we can add a "passive" or "active" to the NfcMessageEvent structure. 

watch sets up a watcher for incoming devices and onread will be called if the WatchOption filter fits. Multiple can be set and the order set in the browser is what matters (if other sites does the same in between).

As chrome books etc can have multiple devices (ie added using USB) I added a findAdapter instead which will list adapters you are allowed to access.

@zolkis @marcoscaceres

New examples:

```
nfc.findAdapter().then(function(adapters) {
  var adapter = adapters[0]; // take the first.

  adapter.onmessage = (ev) => {
    var timesRead = (ev.message.length > 0) ? ev.message[0].timesRead : 0;
    adapter.write({ timesRead: timesRead + 1 }, "/exampleapp");
  }
  adapter.watch({ scope: "/exampleapp"
    }).then(function() {
      console.log("We are now watching!");
    }).catch(function(err) {
      console.log("Error: " + err);
    });
  );
).catch(err) {
  console.log("Couldn't find or access any NFC device.");
}
```

```
adapter.setPushMessage(message, { scope: "/exampleapp"
  }).then(function() {
    watcher.onmessagepushed = () => alert("We successfully pushed a message");
  }).catch(function(err) {
    console.log("Error: " + err);
  });
);
```

**Comment by zolkis** [comment]
adding @don

**Comment by kenchris** [comment]
We should probably decide on this before we publish. I can work on changing the spec tomorrow if this is fine with @zolkis

**Comment by zolkis** [comment]
The spec is already in CfC, the process has to go through, get comments, issues, etc. We can have this issue separately and change the then public draft afterwards. I would like to hear feedback and ideas on web-ish NFC from others in the group.

**Comment by zolkis** [comment]
@kenchris could you explain how do you imagine .write() working?

**Comment by kenchris** [comment]
@zolkis I didn't write out all the arguments. write() should basically work like it does today in our spec.

**Comment by zolkis** [comment]
We need to work on .write() a bit more. It could involve a user interaction (for e.g. "tap now"), whatever the platform needs to do, so from the implementation pov we could just have a timeout for that user interaction, then resolve or reject the promise. We don't need to change the API, but it would be important to mention how implementations are supposed to handle writes.

**Comment by kenchris** [comment]
Sure, we might even give the timeout as part of the arguments to the method. But would that be any different that myself polling for say 3 seconds?

**Comment by don** [comment]
I really like the idea of eliminating tag vs peer for incoming messages. NFC data will be read by the OS, and shows up as an event in the app. The app doesn't have control over when this data is read, it just shows up after the tag is scanned. It doesn't matter if the NDEF message is from a tag or a peer. 

If the message is from a tag, the NfcMessageEvent could contain tag meta when it's available. Meta data could include tagId, techType, nfcForumType, tagCapacity, isFormatted, isReadOnly. Android provides this meta data. Windows Phone does not. On WP8 I'm not sure you can determine if the message is from a peer or tag.

**Comment by don** [comment]
For `write`, Android and WP8 handle this differently. Android expects you to call write when a NFC tag is in range. WP8 queues the data to be written until a tag is in range. Conceptually I like the WP8 way better, since you can call `write` outside of `onread`. I'd like to see clarification on when `write` can be called.

Adding a `cancelWrite` method will allow the user to control the write timeout.

For 9.2.7 I'd like to see "WriteFailedError" rather than "NotSupportedError". Optional additional info for why it failed would be nice.

**Comment by zolkis** [comment]
We have a problem with this unification of tags and peers. It appears that for read, yes, the unification is possible and desired. But for send/write, the situation is a lot more complex, and problematic.
Imagine we call setPushMessage(msg, scope). If it was applied to tags as well, then we say “at any touch event, attempt writing or sending this information with this scope”. But then when a writeable tag is touched, it is automatically overwritten with information not necessarily meant for that tag. For peers it is fine.
Therefore we need a separate write interface for tags. First we need to read a tag, get the read event, then the app can tell its scope, whether it’s writeable etc, and decide it wants to do a write. If the user breaks the range, it will fail (likely the app will ask the user to touch again, or hold touching to start with).
The current interface (separate peer and tag interfaces) was better in the sense that one could set up parallel read+write behaviour, each per scope, and per tag and peer. This was clear and concise, and it is not possible with the current proposal. Plus, it mixes tag behaviour (state machine) with peer behaviour. 
I suggest we save this work as an experimental branch and leave the original proposal in place.

**Comment by don** [comment]
You're correct about writing, it only works for tags and share only sends to peers. I like the idea of keeping these separate.

I think it's more helpful to think of `read` as an event rather than a function. The application registers or subscribes to be notified about NDEF messages. When a tag or peer is scanned, the OS sends an event to the app. It doesn't matter if that message was from a peer or tag, the data is handled the same way.

On Android the tag can be read before it is written, so you can check isWritable, capacity etc. On Windows Phone, you just call proximityDevice.PublishBinaryMessage(...) and the tag is automatically written when it comes in range. IIRC BlackBerry is similar to Android.

**Comment by zolkis** [comment]
On Windows, you can control whether it is meant for tag through the `messageType` parameter, right?
So we can keep a separate write interface for peers and tags.
If we unify the reads, since we need the adapter concept, we arrive to the proposed solution. If you find it good enough, we can go for it. Please check:

```
partial interface Navigator {
    readonly    attribute Nfc nfc;
};

interface Nfc : EventHandler {
  Promise<sequence<NfcAdapter>> findAdapters();
}

typedef NdefData = sequence<URL | Blob | JSON | DOMString>;

interface NfcAdapter : EventHandler {
  Promise<long> watch(WatchOptions? options);  // by default nothing is watched
  Promise clearWatch(long id);

  long setPushMessage(NdefData data, optional USVString scope);
  void clearPushMessages(optional long[] ids);  // if array not given, removes all

  Promise write(NdefData data, optional USVString scope);

  attribute EventHandler<NdefMessageEvent>     onread;  // both tag and peer
  attribute EventHandler onpush;  // successful peer push
  attribute EventHandler onbeforepush;  // called when in touch, but before pushing
}

interface NdefMessageEvent {
  readonly attribute USVString scope;
  readonly attribute boolean passive;
  readonly attribute boolean writeable;
  readonly attribute sequence<NdefRecordData> data;
  ArrayBuffer arrayBuffer();
}

interface NdefRecordData {
  readonly attribute DOMString contentType;
  DOMString url();
  ArrayBuffer arrayBuffer();
  Blob blob();
  Any json();
  DOMString text();
}

dictionary WatchOptions {
  USVString scope;  // must be sub-scope of origin (default)
  DOMString acceptedContentType;
}
```

**Comment by zolkis** [comment]
However, I still like more having separate registration objects for peers and tags, per scope. It  provides a nice way to set up NFC usage from apps, potentially different behavior (event handlers) for different scope, and then separately for tags and peers.

**Comment by zolkis** [comment]
This proposal for simplification has been merged in #58, therefore closing the issue.
